redhat-operators/codeready-workspaces/codeready-workspaces.1.1.0.csv.yaml

```diff
@@ -0,0 +1,345 @@
+apiVersion: operators.coreos.com/v1alpha1
+kind: ClusterServiceVersion
+metadata:
+  name: crwoperator.v1.1.0
+  namespace: placeholder
+  annotations:
+    alm-examples: |-
+      [
+        {
+          "apiVersion":"org.eclipse.che/v1",
+          "kind":"CheCluster",
+          "metadata":{
+             "name":"codeready"
+          },
+          "spec":{
+             "server":{
+                "cheFlavor":"codeready",
+                "cheImage": "quay.io/crw/server-container",
+                "cheImageTag": "1.1-45",
+                "tlsSupport":false,
+                "selfSignedCert":false
+             },
+             "database":{
+                "externalDb":false,
+                "chePostgresHostname":"",
+                "chePostgresPort":"",
+                "chePostgresUser":"",
+                "chePostgresPassword":"",
+                "chePostgresDb":""
+             },
+             "auth":{
+                "openShiftoAuth":false,
+                "externalKeycloak":false,
+                "keycloakURL":"",
+                "keycloakRealm":"",
+                "keycloakClientId":""
+             },
+             "storage":{
+                "pvcStrategy":"common",
+                "pvcClaimSize":"1Gi",
+                "preCreateSubPaths": true
+             }
+          }
+        }
+      ]
+    certified: "true"
+    categories: "Developer Tools"
+    containerImage: quay.io/crw/operator-container:1.1-7
+    createdAt: 2019-03-06
+    description: A Kube-native development solution that delivers portable and collaborative developer workspaces in OpenShift.
+    support: Red Hat, Inc.
+    capabilities: Seamless Upgrades
+spec:
+  install:
+    strategy: deployment
+    spec:
+      clusterPermissions:
```

Comment:
This operator supports the single namespace installMode, but uses clusterPermissions here. Are these required to have the operator run?

Reply:
@SamiSousa Good catch. Yes, I need permissions in a single namespace. Fixed.

However, I wonder if you can help me with the best solution for the following scenario.

In certain scenarios, the operator service account may need permissions:

1. To create, get, list and delete oAuthClients at a cluster scope
2. To get secrets in openshift-ingress namespace.

Currently, I just write down those potential pre-reqs in docs, so that an admin can grant such permissions manually:

```
    oc create clusterrole codeready-operator --resource=oauthclients --verb=get,create,delete,update,list,watch

    oc create clusterrolebinding codeready-operator --clusterrole=codeready-operator --serviceaccount=${NAMESPACE}:codeready-operator

    oc create role secret-reader --resource=secrets --verb=get -n=openshift-ingress

    oc create rolebinding codeready-operator --role=secret-reader --serviceaccount=${NAMESPACE}:codeready-operator -n=openshift-ingress
```

Is there a way I can declare such clusterPermissions in a csv?

Reply:
I believe you can have both the `permissions` section and a `clusterPermissions` section. In the `permissions`, you should list all the single namespace permissions for the operator, and in `clusterPermissions` you should list the oAuthClients permissions and the external namespace permissions.
@ecordell Does this sound right?

Reply:
@SamiSousa thanks, this is what I had in mind. I just wonder if it is possible not to let a service account read all secrets in the cluster, but only in a particular namespace (openshift-ingress in my case)?

Reply:
I'm not aware of a way to have service accounts watch particular namespaces, I don't think OLM has that ability without asking users to install this operator as a SingleNamespace install in the `openshift-ingress` namespace.
@ecordell Thoughts?

Reply:
@SamiSousa let me explain choices I have.

1. Security first. When needed a cluster admin manually provisions those roles and rolebindings
2. UX first. OLM creates clusterRole and clusterRolebinding, however, since the operator can be deployed to an individual namespace and supports multi namespace, it means users will have these  superpowers.

So, I am good with having instructions (choice 1) saying what to do if you need to use particular CodeReady Workspaces features. Besides, I I am planning to get rid of reading a tls secret in openshift ingress namespace and fetch router crt in a different way, so it will be just one potentially missing ClusterRole - manage oAuthclients.

Reply:
You can also define required RBAC in `permissions` rather than `clusterPermissions` and OLM will automatically upgrade these to clusterPermissions in an AllNamespaces installMode.

Reply:
@SamiSousa yes, one of the options. We'd prefer however to have a single namespace install mode.

Reply:
@eivantsov Does that mean that the operator needs cluster permissions to run in a singleNamespace installMode?

Reply:
No. With default settings, no. When it does, there are instructions in description about what roles and rolebindings are to be manually created if a user(admin) needs certain operator functionality.

Reply:
There is no additional security benefit by having the cluster-admin manually provision the roles and bindings since he would also be the only one who is able to install an Operator.

Reply:
But with a single namespace install mode the operator can be installed in any cluster. And a new clusterRoleBinding will be created each time.